iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java

```diff
@@ -148,6 +148,14 @@ public void initialize(@Nullable Configuration configuration, Properties serDePr
     // TODO: remove once we have both Fanout and ClusteredWriter available: HIVE-25948
     HiveConf.setIntVar(configuration, HiveConf.ConfVars.HIVEOPTSORTDYNAMICPARTITIONTHRESHOLD, 1);
     HiveConf.setVar(configuration, HiveConf.ConfVars.DYNAMICPARTITIONINGMODE, "nonstrict");
+
+    Context.Operation operation = HiveCustomStorageHandlerUtils.getWriteOperation(configuration,
+            serDeProperties.getProperty(Catalogs.NAME));
+
+    if (operation != null) {
+      HiveConf.setFloatVar(configuration, HiveConf.ConfVars.TEZ_MAX_PARTITION_FACTOR, 1f);
```

Comment:
I tested how this works as I am not confident.

## Test queries

### Prep

```
beeline -e "
DROP TABLE IF EXISTS test;
CREATE TABLE test (id INT) STORED BY ICEBERG TBLPROPERTIES('format-version'='2');
INSERT INTO test VALUES (1), (2), (3), (4), (5), (6), (7), (8), (9), (10);
"
```

### DELETE

```
beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true
```

## Test result

### The original version

[This](https://github.com/okumin/hive/commit/59b84b9d7835f97b5e9df872ef00986678498976) is the tested revision. Two reducers are set up and launched as reported.

```
$ beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true
...
INFO  : 2023-07-16 04:55:39,740	Map 1: 0(+1)/1	Reducer 2: 0/2	
INFO  : 2023-07-16 04:55:42,259	Map 1: 1/1	Reducer 2: 0(+1)/2
```

```
$ ./bin/logs hive-hiveserver2-655b558bb-gzwh8 | grep HIVE-27050 | tail -n 1
hive-hiveserver2-655b558bb-gzwh8: 2023-07-16T04:55:31,760  INFO [f3ecdcb4-f45b-405f-986e-0dbf525c7c87 HiveServer2-Handler-Pool: Thread-61] parse.GenTezUtils: HIVE-27050: max partition factor=2.0, max partition=2
```

### The first patch

[This](https://github.com/okumin/hive/commit/a19ee2b5536539d066ecf77c369bc0f7bc85a2ca) is the tested revision. One reducer is set up and launched.

```
$ beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true
...
INFO  : 2023-07-16 05:22:30,905	Map 1: 0/1	Reducer 2: 0/1	
INFO  : 2023-07-16 05:22:35,454	Map 1: 0(+1)/1	Reducer 2: 0/1
```

```
$ ./bin/logs hive-hiveserver2-75bc6bc94c-bhxxc | grep HIVE-27050 | tail -n 1
hive-hiveserver2-75bc6bc94c-bhxxc: 2023-07-16T05:22:23,324  INFO [18f9ace5-2016-4430-8155-ae95c254148b HiveServer2-Handler-Pool: Thread-61] parse.GenTezUtils: HIVE-27050: max partition factor=1.0, max partition=1
```

### The second(current) path

[This](https://github.com/okumin/hive/commit/0e149323aa07b661a0c8eb259fa514bcb72c02a9) is the tested revision. It is unlikely to be working as expected.

```
$ beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true
...
INFO  : 2023-07-16 05:58:27,235	Map 1: 1/1	Reducer 2: 0/2	
INFO  : 2023-07-16 05:58:27,741	Map 1: 1/1	Reducer 2: 0(+1)/2
```

```
$ ./bin/logs hive-hiveserver2-5876b658cd-8zxwp | grep HIVE-27050 | tail -n 3
hive-hiveserver2-5876b658cd-8zxwp: 2023-07-16T05:58:13,110  INFO [62373162-813c-409b-834d-335f2f72efc1 HiveServer2-Handler-Pool: Thread-61] parse.GenTezUtils: HIVE-27050: max partition factor=2.0, max partition=2
hive-hiveserver2-5876b658cd-8zxwp: 2023-07-16T05:58:13,112  INFO [62373162-813c-409b-834d-335f2f72efc1 HiveServer2-Handler-Pool: Thread-61] hive.HiveIcebergSerDe: HIVE-27050: operation: null
hive-hiveserver2-5876b658cd-8zxwp: 2023-07-16T05:58:28,411  INFO [HiveServer2-Background-Pool: Thread-111] hive.HiveIcebergSerDe: HIVE-27050: operation: null
```

Reply:
Some random thoughts. I would say these are minor.
- Is it best to disable over-provisioning only for DELETE reducers using a Hive hook or something?
    - Over-provisioning might work for INSERT or UPDATE
    - It could work also for DELETE if rows are filtered
    - Maybe, the optimization should be applied to only the last reducer in the case like Map -> Reduce -> Reduce?
- `hive.tez.auto.reducer.parallelism.min.threshold=0.0` can be an option?

Reply:
Hi @okumin, 
Thank you for all the info!
I tried a DELETE query with your test table with `hive.tez.auto.reducer.parallelism.min.threshold=0.0 ` and it still gave 2 reducers.
I also noticed in debug that `GenTezUtils.createReduceWork()` which sets max partition factor is called before `HiveIcebergSerDe.initialize()` with a write operation where I tried to set max partition factor to 1 which probably explains why my change doesn't work.
What do you think, can I set max partition factor to 1.0 in `GenTezUtils.createReduceWork()` if the operation is a write operation for Iceberg table?

Reply:
I personally think it is reasonable to explicitly inject the logic into GenTezUtils or somewhere. One request is I'd like to make it pluggable because other formats would hit the same issue. As far as I checked on my machine, Hive ACID shares the problem. Note that these are my opinions and committers could have different ideas, or they might think it is an expected behavior.

As for the parameter, I guess you tested it with 4.0.0-alpha-2 since the param was merged recently. It enforces auto reduce parallelism.

```
$ beeline -e "
> DROP TABLE IF EXISTS test;
> CREATE TABLE test (id INT) STORED BY ICEBERG TBLPROPERTIES('format-version'='2');
> INSERT INTO test VALUES (1), (2), (3), (4), (5), (6), (7), (8), (9), (10);
> "
```

```
$ beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true
...
INFO  : 2023-07-18 14:19:45,200	Map 1: 0(+1)/1	Reducer 2: 0/2	
INFO  : 2023-07-18 14:19:48,227	Map 1: 1/1	Reducer 2: 0(+1)/2
...
$ beeline -e 'DELETE FROM test WHERE id = 5' --hiveconf hive.server2.in.place.progress=false --hiveconf hive.tez.auto.reducer.parallelism=true --hiveconf hive.tez.auto.reducer.parallelism.min.threshold=0.0
...
INFO  : 2023-07-18 14:20:23,730	Map 1: 0(+1)/1	Reducer 2: 0/2	
INFO  : 2023-07-18 14:20:27,271	Map 1: 1/1	Reducer 2: 0(+1)/1
```

Reply:
I updated the solution and placed it in GenTezUtils and made it for Iceberg and Hive ACID.